provider/provider.go

```diff
@@ -0,0 +1,165 @@
+package provider
+
+import (
+	"context"
+	"gx/ipfs/QmR8BauakNcBa3RbE4nbQu76PDiJgoQgz8AJdhJuiU4TAw/go-cid"
+	"gx/ipfs/QmZBH87CAPFHcc7cYmBqeSQ98zQ3SX9KUxiYgzPmLWNVKz/go-libp2p-routing"
+	logging "gx/ipfs/QmcuXC5cxs79ro2cUuHs4HQ2bkDLJUYokwL8aivcX6HW3C/go-log"
+	"sync"
+	"time"
+)
+
+var (
+	log = logging.Logger("provider")
+)
+
+const (
+	provideOutgoingWorkerLimit = 512
+	provideOutgoingTimeout     = 15 * time.Second
+)
+
+type Strategy func(context.Context, cid.Cid) <-chan cid.Cid
+
+type Provider struct {
+	ctx context.Context
+	lock sync.Mutex
+
+	// CIDs to provide
+	incoming chan cid.Cid
+	// CIDs we are working on providing now
+	outgoing chan cid.Cid
+	// strategy for deciding which CIDs, given a CID, should be provided
+	strategy Strategy
+	// keeps track of which CIDs have been provided already
+	tracker *Tracker
+	// the CIDs for which provide announcements should be made
+	queue *Queue
+	// used to announce providing to the network
+	contentRouting routing.ContentRouting
+}
+
+func NewProvider(ctx context.Context, strategy Strategy, tracker *Tracker, queue *Queue, contentRouting routing.ContentRouting) *Provider {
+	return &Provider{
+		ctx:            ctx,
+		strategy:       strategy,
+		tracker:        tracker,
+		queue:          queue,
+		contentRouting: contentRouting,
+		lock:           sync.Mutex{},
+		outgoing:       make(chan cid.Cid),
+		incoming:       make(chan cid.Cid),
+	}
+}
+
+// Start workers to handle provide requests.
+func (p *Provider) Run() {
+	go p.handleIncoming()
+	go p.handlePopulateOutgoing()
+	go p.handleOutgoing()
+}
+
+// Provider the given cid using specified strategy.
+func (p *Provider) Provide(root cid.Cid) {
+	cids := p.strategy(p.ctx, root)
```

Comment:
This makes sense. I'm going to address this feedback last so we can discuss it in more detail after I get the other feedback incorporated. Briefly, though, my thoughts are:

Would we still need to allow a "global" providing strategy to be configured? How does that work with the ability to call provider strategies explicitly? Would we need to use an abstraction at all of the provide call sites that chooses the actual strategy to use based on the config?

Reply:
> Would we need to use an abstraction at all of the provide call sites that chooses the actual strategy to use based on the config?

Yes. The issue is that we'll likely want different default strategies for different commands (files, other data, pinning, etc.). For example, a "provide every directory/file root" strategy won't make sense as a default for the `ipfs dag` or `ipfs block` commands because *those* work directly with IPLD blocks (which may not be files).

However, it may make sense to introduce *another* service for this. Maybe? That is, something with a function like `GetStrategyFor(stuff) (Strategy, error)` where `stuff` includes things like:

* Whether or not we're pinning.
* The type of data we're handling (raw blocks, ipfs files, etc.).
* Maybe even the root CID?
* Maybe even the current command?

Really, I'm not sure what the right design is here.

Reply:
Understood, I'm processing this now.

Reply:
@Stebalien It sounds like the plan is to move the provide calls to the call-sites and to choose an appropriate providing strategy for each location. However, there is currently the ability to configure a providing strategy, which we've expressed the desire to preserve. I'm trying to figure out exactly what those configurations would do.

I think a reasonable approach is to determine all the places where a "provide all" currently takes place (e.g. `ipfs add`) and to allow those locations to be configured by the user's provider strategies. This makes the system behave a little different from the way it behaves right now, which I think is unavoidable, but achieves the larger goal of allowing a limited strategy to be used in place where there's the potential to provide too many blocks.

Any thoughts?

Reply:
Just to clarify, in the locations where a "provide all" does not make sense, we just always use whatever providing strategy we decide makes the most sense in that location.

Reply:
> I think a reasonable approach is to determine all the places where a "provide all" currently takes place (e.g. ipfs add) and to allow those locations to be configured by the user's provider strategies. 

Exactly.

> This makes the system behave a little different from the way it behaves right now, which I think is unavoidable

Correct. Really, we can probably deprecate the current reprovider strategy system and write a config migration.

---

For example, we could (**eventually**) end up supporting a provider strategy spec like the following:

```js
{
  // Default strategy.
  "default": {
    // all objects.
    "all": {
      "strategy": "root",
      // put it in the queue, announce once, but don't record or reprovide
      // We probably won't support this out of the box...
      "when": "once",
    }
  }
  // For `ipfs add --pin=false`. That is, data _we've_ chosen to add (but not pin).
  "add": {
    "all": {
      "strategy": "root",
    },
    "unixfs": {
      "strategy": "files",
      "when": "once",
    }
  },
  // Either `ipfs pin /ipfs/...` or `ipfs add ...`
  "pin": {
    // fallback.
    // In this case, it only applies ot `ipfs pin /ipld/Qm...`
    "default": {
      "strategy": "all",
    },
    // Applies to things like `ipfs pin /ipfs/Qm...`
    "unixfs": {
      "strategy": "files",
    },
  },
}
```

(however, the user should be able to override the strategy when adding/pinning data.